File: TrappingRainWater.java
import java.util.Scanner;

public class TrappingRainWater {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter length of array:");
        int len = sc.nextInt();

        int[] bars = new int[len];
        for (int i = 0; i < bars.length; i++) {
            bars[i] = sc.nextInt();
        }
        sc.close();
        
        int quantity = method1(bars);
        System.out.println(quantity);
    }

    private static int method1(int[] arr) {
        int left = 0, right = arr.length - 1; // 'left' pointer points to first index initially and 'right' pointer
                                              // points to the last element in the array initially

        int maxLeft = 0, maxRight = 0, ans = 0;// maxLeft stores height of longest bar on the left of a certain bar
                                               // maxRight stores the height of longest bar on the right of a certain
                                               // bar
        while (left < right) { // run the loop as long as left pointer doesn't cross right pointer

            if (arr[left] < arr[right]) { // if arr[left] < arr[right], update maxLeft and ans and increment the left
                                          // pointer
                maxLeft = Math.max(maxLeft, arr[left]);
                ans += maxLeft - arr[left];
                left++;
            }

            else { // else, perform the same steps but in this case, decrement right pointer
                maxRight = Math.max(maxRight, arr[right]);
                ans += maxRight - arr[right];
                right--;
            }
        }

        return ans;
    }
}

File: trapping_rain_water.cpp
#include<bits/stdc++.h>

using namespace std;


class Solution{

   
    public:
    long long trappingWater(int arr[], int n){
        
        int left[n];
     int right[n];

     left[0] = arr[0];
     for(int i=1; i<n; i++){
         left[i] = max(arr[i], left[i-1]);
     }

     right[n-1] = arr[n-1];
     for(int i=n-2; i>=0; i--){
         right[i] = max(right[i+1] , arr[i]);
     }

     long ans = 0;
     for(int i=0; i<n; i++){
         ans =ans + (min(left[i] , right[i])) - arr[i];
     }
     return ans;
    }
};



int main(){

    int t;
    //testcases
    cin >> t;

    while(t--){
        int n;


        cin >> n;

        int a[n];


        for(int i =0;i<n;i++){
            cin >> a[i];
        }
        Solution obj;

        cout << obj.trappingWater(a, n) << endl;

    }

    return 0;
}
